atef/tools.py

```diff
@@ -0,0 +1,200 @@
+    #: Host name to time taken.
+    times: Dict[str, float]
+    #: Minimum time from ``times``.
+    min_time: float
+    #: Maximum time from ``times``.
+    max_time: float
+
+
+@dataclasses.dataclass
+@serialization.as_tagged_union
+class Tool:
+    """
+    Base class for atef tool checks.
+    """
+    result_type: ClassVar[Type[ToolResult]] = ToolResult
+
+    def check_result_key(self, key: str) -> None:
+        valid_keys = list(typing.get_type_hints(self.result_type))
+        if key not in valid_keys:
+            raise KeyError(
+                f"Invalid result key for tool {self}: {key!r}.  Valid "
+                f"keys are: {', '.join(valid_keys)}"
+            )
+
+    async def run(self, *args, **kwargs) -> ToolResult:
+        raise NotImplementedError("To be implemented by subclass")
+
+
+@dataclasses.dataclass
+class Ping(Tool):
+    """
+    Tool for pinging one or more hosts and summarizing the results.
+    """
+    #: The hosts to ping.
+    hosts: List[str] = field(default_factory=list)
+    #: The number of ping attempts to make per host.
+    count: int = 3
+    #: The assumed output encoding of the 'ping' command.
+    encoding: str = "utf-8"
+
+    result_type: ClassVar[Type[ToolResult]] = PingResult
+    _time_re: ClassVar[re.Pattern] = re.compile(r"time=(.*)\s?ms")
+    _unresponsive_time: ClassVar[float] = 100.0
+
+    @staticmethod
+    def _result_from_output(host: str, output: str) -> PingResult:
+        """
+        Fill a PingResult from the results of the ping program.
+
+        Parameters
+        ----------
+        host : str
+            The hostname that ``ping`` was called with.
+        output : str
+            The decoded output of the subprocess call.
+
+        Returns
+        -------
+        PingResult
+        """
+        # NOTE: lazily ignoring non-millisecond-level results here; 1 second+
+        # is the same as non-responsive if you ask me...
+        times = [float(ms) for ms in Ping._time_re.findall(output)]
+
+        if not times:
+            return PingResult(
+                result=Result(severity=Severity.error),
+                alive=[],
+                unresponsive=[host],
+                min_time=Ping._unresponsive_time,
+                max_time=Ping._unresponsive_time,
+                times={host: Ping._unresponsive_time},
+            )
+
+        return PingResult(
+            result=Result(severity=Severity.success),
+            alive=[host],
+            unresponsive=[],
+            min_time=min(times),
+            max_time=max(times),
+            times={host: sum(times) / len(times)},
+        )
+
+    async def ping(self, host: str) -> PingResult:
+        """
+        Ping the given host.
+
+        Parameters
+        ----------
+        host : str
+            The host to ping.
+
+        Returns
+        -------
+        PingResult
+        """
+
+        if sys.platform == "win32":
+            args = ("/n", str(self.count))
```

Comment:
Untested on Windows

Reply:
My windows machine has nothing useful to ping except maybe I can make a config that pings google

Reply:
`localhost` will work just fine (see the test suite + included config file)

Reply:
Would it be helpful then to simply run the test suite on my windows machine?

Reply:
If you wouldn't mind, that could help. I doubt we'll ever get a CI job for Windows configured.
Alternatively, if you want to run `atef check -vv atef/tests/configs/ping_localhost.json` that'll do, too.

Reply:
Mixed results here: the test suite fails and the atef check passes.
```
_____________________________________________________ test_result_severity[all_good] ______________________________________________________ 

conf = ToolConfiguration(name=None, description=None, tags=None, checklist=[IdentifierAndComparison(name=None, ids=['max_time...27.0.0.1'], 
times={'127.0.0.1': 100.0}, min_time=0.0, max_time=100.0), hosts=['127.0.0.1'], count=1, encoding='utf-8'))
severity = <Severity.success: 0>

    @config_and_severity
    @pytest.mark.asyncio
    async def test_result_severity(
        conf: ToolConfiguration, severity: Severity
    ):
        overall, results = await config.check_tool(conf.tool, conf.checklist)
>       assert overall == severity
E       assert <Severity.error: 2> == <Severity.success: 0>

atef\tests\test_comparison_tools.py:61: AssertionError
```

Reply:
Hmm, running something like `pytest --pdb --lf`, what's in the `results` list?

Reply:
```
[Result(severity=<Severity.error: 2>, reason='max_time <= 1: value of 100.0')]
```
Does windows ping... have a different time scale?

Reply:
I'm a bit confused but I'll look a little more closely
```
C:\Users\zlentz>ping localhost

Pinging PC98346.win.slac.stanford.edu [127.0.0.1] with 32 bytes of data:
Reply from 127.0.0.1: bytes=32 time<1ms TTL=128
Reply from 127.0.0.1: bytes=32 time<1ms TTL=128
Reply from 127.0.0.1: bytes=32 time<1ms TTL=128
Reply from 127.0.0.1: bytes=32 time<1ms TTL=128

Ping statistics for 127.0.0.1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 0ms, Average = 0ms
```

Reply:
Odd, I think the regular expression didn't match any times in `ping` in that case. The 100 is the default value to use on error (separate bug that reminded me of: need to normalize times in seconds instead of the awkward `ms`)

Reply:
I think the problem is that the output from windows is formatted completely differently

Reply:
Oh, did it by chance switch units to microseconds/nanoseconds? I thought I recalled a platform that did that, but neither macOS nor Linux were deviating from `ms`

Reply:
I'll switch up the regular expression to match `<1ms` as just `1ms`. I hope that'll be good enough.